astropy/io/fits/convenience.py

```diff
@@ -186,22 +186,34 @@ def getdata(filename, *args, header=None, lower=None, upper=None, view=None,
 
         If the optional keyword ``header`` is set to `True`, this
         function will return a (``data``, ``header``) tuple.
+
+    Raises
+    ------
+    ValueError
+        If HDU has no data.
     """
 
     mode, closed = _get_file_mode(filename)
 
+    ext = kwargs.get('ext')
+    extname = kwargs.get('extname')
+    extver = kwargs.get('extver')
+    no_user_ext = (len(args) == 0 and ext is None and
+                   extname is None and extver is None)
+
     hdulist, extidx = _getext(filename, mode, *args, **kwargs)
     try:
         hdu = hdulist[extidx]
         data = hdu.data
-        if data is None and extidx == 0:
+        if data is None and no_user_ext:
+            # fallback to the first non-primary extension
             try:
                 hdu = hdulist[1]
                 data = hdu.data
             except IndexError:
-                raise IndexError('No data in this HDU.')
+                pass
```

Comment:
```suggestion
                raise IndexError('No data in primary HDU and no extensions found.')
```
The discussion about which exception type to expose to the user aside, that one actually is a genuine `IndexError`, but was giving incorrect diagnostic. I think this still deserves to be caught separately to remind the user that `getdata` was looking in the default locations.